Added a ROS Transform Publisher and an Example Car With LIDAR Sensor Simulation

liangfok opened the pull request:
Closes https://github.com/RobotLocomotion/drake/issues/2211

To run the new simulation, ensure `[drake-distro]/drake/examples/Cars/ros_packages` is in your `ROS_PACKAGE_PATH`. Then execute the following commands:

```
$ roslaunch drake_cars_examples rviz_prius.launch
$ cd [drake distro]/drake/examples/Cars
$ ../../pod-build/bin/carSimLCMandROS models/prius/prius_with_lidar.sdf models/stata_garage_p1.sdf
```
The above instructions are captured in the updated `[drake distro]/drake/examples/Cars/README.md`.

You should see something like:

![screenshot from 2016-05-03 16 21 19](https://cloud.githubusercontent.com/assets/1388098/14997341/4ffb0420-114c-11e6-9af0-85d0181844da.png)

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/robotlocomotion/drake/2253) &emsp; Multiple assignees:&emsp;<img alt="@david-german-tri" height="20" width="20" align="absmiddle" src="https://avatars.githubusercontent.com/u/17437069?s=40&v=3">&nbsp;<a href="/robotlocomotion/drake/pulls/assigned/david-german-tri">david-german-tri</a>,&emsp;<img alt="@jwnimmer-tri" height="20" width="20" align="absmiddle" src="https://avatars.githubusercontent.com/u/17596505?s=40&v=3">&nbsp;<a href="/robotlocomotion/drake/pulls/assigned/jwnimmer-tri">jwnimmer-tri</a>,&emsp;<img alt="@liangfok" height="20" width="20" align="absmiddle" src="https://avatars.githubusercontent.com/u/1388098?s=40&v=3">&nbsp;<a href="/robotlocomotion/drake/pulls/assigned/liangfok">liangfok</a>,&emsp;<img alt="@nkoenig" height="20" width="20" align="absmiddle" src="https://avatars.githubusercontent.com/u/1587438?s=40&v=3">&nbsp;<a href="/robotlocomotion/drake/pulls/assigned/nkoenig">nkoenig</a>
<!-- Reviewable:end -->

liangfok commented:
The following cpplint issues were reported:

```
16:35:42 /home/ubuntu/workspace/experimental/ae7636ba/drake/systems/plants/drake_ros_tf_publisher.h:60: warning: Single-parameter constructors should be marked explicit.  [runtime/explicit] [5]
16:35:44 /home/ubuntu/workspace/experimental/ae7636ba/drake/systems/plants/sensor_visualizer_lidar.h:53: warning: Single-parameter constructors should be marked explicit.  [runtime/explicit] [5]
16:35:49 /home/ubuntu/workspace/experimental/ae7636ba/drake/examples/Cars/simulateLCMandROS.cpp:26: warning: Single-parameter constructors should be marked explicit.  [runtime/explicit] [5]
```

I shall stop the CI build. Fix the above errors. And then restart a new CI Build.

liangfok commented:
The Win64 flake in [CI Build 1207](https://drake-jenkins.csail.mit.edu/job/experimental/1207/) appears to be related to the problems in Win32 here: https://drake-jenkins.csail.mit.edu/job/windows-msvc-32-continuous-debug/

liangfok commented:
Since [CI Experimental Build 1219](https://drake-jenkins.csail.mit.edu/job/experimental/1219/) appears to have failed due to the flake tracked by https://github.com/RobotLocomotion/drake/issues/2257, I'm going to elevate this PR to the feature reviewer. @david-german-tri, can you serve as the feature reviewer or assign it to someone who can be one?

liangfok commented:
[CI Experimental Build 1223](https://drake-jenkins.csail.mit.edu/job/experimental/1223/) passed. Thus, this PR is definitely ready for feature review.

david-german-tri commented:
I'm not going to have time to do a detailed review until tomorrow, but here are some high-level comments.

---

Review status: 0 of 27 files reviewed at latest revision, 7 unresolved discussions.

---

*[drake/examples/Cars/CMakeLists.txt, line 27 \[r12\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2253#-KGxnKCshIlnStj6quhH:-KGxnKCshIlnStj6quhI:1693743548) ([raw file](https://github.com/robotlocomotion/drake/blob/7c222f12089e672e34da22832fbd1b6c09c1c278/drake/examples/Cars/CMakeLists.txt#L27)):*
please delete

---

*[drake/examples/Cars/CMakeLists.txt, line 35 \[r12\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2253#-KGxnOXWi60ccVB3pl--:-KGxnOXWi60ccVB3pl-0:2131214325) ([raw file](https://github.com/robotlocomotion/drake/blob/7c222f12089e672e34da22832fbd1b6c09c1c278/drake/examples/Cars/CMakeLists.txt#L35)):*
delete this too, I guess?

---

*[drake/examples/Cars/simulateLCMandROS.cpp, line 1 \[r12\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2253#-KGxoXvlb_0j8r1Gs0iU:-KGxoXvlb_0j8r1Gs0iV:1295201897) ([raw file](https://github.com/robotlocomotion/drake/blob/7c222f12089e672e34da22832fbd1b6c09c1c278/drake/examples/Cars/simulateLCMandROS.cpp#L1)):*
Instead of forking the file, would it make sense to factor out the share code?  (We may have chatted about this before, apologies if I've forgotten.)

---

*[drake/systems/plants/drake_ros_tf_publisher.h, line 1 \[r12\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2253#-KGxogLp6VnbWoCXCR17:-KGxogLp6VnbWoCXCR18:1982965569) ([raw file](https://github.com/robotlocomotion/drake/blob/7c222f12089e672e34da22832fbd1b6c09c1c278/drake/systems/plants/drake_ros_tf_publisher.h#L1)):*
Does this have test coverage?

---

*[drake/systems/plants/drake_ros_tf_publisher.h, line 26 \[r12\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2253#-KGxotAK5-exs1kdBu_U:-KGxotAL1-1IFajMUtkx:-1439627345) ([raw file](https://github.com/robotlocomotion/drake/blob/7c222f12089e672e34da22832fbd1b6c09c1c278/drake/systems/plants/drake_ros_tf_publisher.h#L26)):*
Link explaining the significance of `/tf`?

---

*[drake/systems/plants/sensor_visualizer_lidar.h, line 1 \[r12\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2253#-KGxq09L_Hftf8VsPcBW:-KGxq09L_Hftf8VsPcBX:-1546366242) ([raw file](https://github.com/robotlocomotion/drake/blob/7c222f12089e672e34da22832fbd1b6c09c1c278/drake/systems/plants/sensor_visualizer_lidar.h#L1)):*
There's a lot to digest in this PR.  Would it make sense to put one of the publishers in a separate PR?

---

*[drake/systems/plants/sensor_visualizer_lidar.h, line 1 \[r12\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2253#-KGxr1bSbLu1xKdA44Mv:-KGxr1bSbLu1xKdA44Mw:-484780626) ([raw file](https://github.com/robotlocomotion/drake/blob/7c222f12089e672e34da22832fbd1b6c09c1c278/drake/systems/plants/sensor_visualizer_lidar.h#L1)):*
Test coverage?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2253#-:-KGxr71GvFTINNcDFIo4:1170929563)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Review status: 0 of 27 files reviewed at latest revision, 7 unresolved discussions.

---

*[drake/examples/Cars/CMakeLists.txt, line 27 \[r12\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2253#-KGxnKCshIlnStj6quhH:-KGyDftUqjXOqZp0ifa0:-499146009) ([raw file](https://github.com/robotlocomotion/drake/blob/7c222f12089e672e34da22832fbd1b6c09c1c278/drake/examples/Cars/CMakeLists.txt#L27)):*
Done.

---

*[drake/examples/Cars/CMakeLists.txt, line 35 \[r12\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2253#-KGxnOXWi60ccVB3pl--:-KGyDedUewRYOCH1pdg1:-1684884930) ([raw file](https://github.com/robotlocomotion/drake/blob/7c222f12089e672e34da22832fbd1b6c09c1c278/drake/examples/Cars/CMakeLists.txt#L35)):*
Done. I'll put this line back if / when we start to use custom ROS messages.

---

*[drake/examples/Cars/simulateLCMandROS.cpp, line 1 \[r12\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2253#-KGxoXvlb_0j8r1Gs0iU:-KGyDoHKG7tyxiTBThhl:348758612) ([raw file](https://github.com/robotlocomotion/drake/blob/7c222f12089e672e34da22832fbd1b6c09c1c278/drake/examples/Cars/simulateLCMandROS.cpp#L1)):*
Yes, factoring out the shared code would be ideal. Would you be OK punting on this if I create an issue to track its completion? One reason for punting is because I'm actually making significant changes to this file in my latest efforts at replacing the LCM driving command receiver with a ROS driving command receiver.

---

*[drake/systems/plants/drake_ros_tf_publisher.h, line 1 \[r12\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2253#-KGxogLp6VnbWoCXCR17:-KGyEvpuNp7jrmiNmhjd:-296149109) ([raw file](https://github.com/robotlocomotion/drake/blob/7c222f12089e672e34da22832fbd1b6c09c1c278/drake/systems/plants/drake_ros_tf_publisher.h#L1)):*
It does not. I will add one using `rostest` shortly.

---

*[drake/systems/plants/drake_ros_tf_publisher.h, line 26 \[r12\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2253#-KGxotAK5-exs1kdBu_U:-KGyFJ5ZDVrWvyGHpykS:-499146009) ([raw file](https://github.com/robotlocomotion/drake/blob/7c222f12089e672e34da22832fbd1b6c09c1c278/drake/systems/plants/drake_ros_tf_publisher.h#L26)):*
Done.

---

*[drake/systems/plants/sensor_visualizer_lidar.h, line 1 \[r12\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2253#-KGxq09L_Hftf8VsPcBW:-KGyFL4Vxg5RwATkhfkZ:-1482726355) ([raw file](https://github.com/robotlocomotion/drake/blob/7c222f12089e672e34da22832fbd1b6c09c1c278/drake/systems/plants/sensor_visualizer_lidar.h#L1)):*
I originally had PR https://github.com/RobotLocomotion/drake/pull/2126 but it was stalled for several days waiting for a feature review . During that time, this PR superseded it and it became a chore for me to maintain both branches. So I closed that PR and kept this one open. Should I not do that?

---

*[drake/systems/plants/sensor_visualizer_lidar.h, line 1 \[r12\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2253#-KGxr1bSbLu1xKdA44Mv:-KGyG2Euok4yTIc3tY8k:1774158373) ([raw file](https://github.com/robotlocomotion/drake/blob/7c222f12089e672e34da22832fbd1b6c09c1c278/drake/systems/plants/sensor_visualizer_lidar.h#L1)):*
Will add it soon. Note even with test coverage, CI will not perform test coverage until https://github.com/RobotLocomotion/drake/issues/2116 is resolved.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2253)*
<!-- Sent from Reviewable.io -->

liangfok commented:
@david-german-tri: Let me address some more of your comments before having this PR reviewed again. I'll ping you when ready.

---

Review status: 0 of 27 files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2253#-:-KH5hw-S3g6j0p-Yzslh:-1591575827)*
<!-- Sent from Reviewable.io -->

liangfok commented:
I will close this PR until all conflicts, etc., are resolved.

liangfok commented:
TODO:

- [ ] Only put Drake <-> ROS stuff in `drake/ros` and `drake/examples/**`.

liangfok commented:
+@david-german-tri for feature review since he already reviewed it once.
+@nkoenig for feature review since he should know what we're doing about ROS + Drake integration
+@jwnimmer-tri for platform review once the feature reviews are done.

<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> TODO:
> 
> 
> 
> - [x] Only put Drake <-> ROS stuff in `drake/ros` and `drake/examples/**`.</details>


---

Review status: 0 of 33 files reviewed at latest revision, 7 unresolved discussions.

---

*[drake/examples/Cars/car_sim_lcm_and_ros.cpp, line 1 \[r12\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2253#-KGxoXvlb_0j8r1Gs0iU:-KHMEbLUqn6zzwLCyidp:150439124) ([raw file](https://github.com/robotlocomotion/drake/blob/7c222f12089e672e34da22832fbd1b6c09c1c278/drake/examples/Cars/car_sim_lcm_and_ros.cpp#L1)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> Yes, factoring out the shared code would be ideal. Would you be OK punting on this if I create an issue to track its completion? One reason for punting is because I'm actually making significant changes to this file in my latest efforts at replacing the LCM driving command receiver with a ROS driving command receiver.</details>
Done (I refactored the code to pull out common code).

---

*[drake/systems/plants/drake_ros_tf_publisher.h, line 1 \[r12\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2253#-KGxogLp6VnbWoCXCR17:-KHMTGzNtFZdQtsuS2Ti:-1472198992) ([raw file](https://github.com/robotlocomotion/drake/blob/7c222f12089e672e34da22832fbd1b6c09c1c278/drake/systems/plants/drake_ros_tf_publisher.h#L1)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> It does not. I will add one using `rostest` shortly.</details>
I added unit test `car_sim_lcm_and_ros`, which covers this new system. To run the unit test:

```
$ ctest -VV -R car_sim_lcm_and_ros
```

---

*[drake/systems/plants/sensor_visualizer_lidar.h, line 1 \[r12\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2253#-KGxr1bSbLu1xKdA44Mv:-KHMTSRLB9_2TUtbYriY:-1472198992) ([raw file](https://github.com/robotlocomotion/drake/blob/7c222f12089e672e34da22832fbd1b6c09c1c278/drake/systems/plants/sensor_visualizer_lidar.h#L1)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> Will add it soon. Note even with test coverage, CI will not perform test coverage until https://github.com/RobotLocomotion/drake/issues/2116 is resolved.</details>
I added unit test `car_sim_lcm_and_ros`, which covers this new system. To run the unit test:

```
$ ctest -VV -R car_sim_lcm_and_ros
```

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2253#-:-KHMTgBhyKwdoOYzTaaU:-398553693)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
@liangfok 

> +@jwnimmer-tri for platform review once the feature reviews are done.

Sure!  It sounds like I should wait to review until others do a first pass.  This PR is large, so expect it might take me ~2 days of latency to get fully through once I start.  I might still do a design review of `car.h` up front, because it will affect my own work in the near term.

Would you like to consider breaking out the RB/RBT/RBS changes into their own PR, so we can iterate those out first, and more quickly?  They seem fairly self-contained.

liangfok commented:
I would be happy to pull out the changes to `RigidBody`, `RigidBodyTree`, and `RigidBodySystem` into a separate PR. However, I'm fairly sure these modifications will have conflicts with https://github.com/RobotLocomotion/drake/pull/2207. Thus, I think I should wait for https://github.com/RobotLocomotion/drake/pull/2207 to be merged, fix the conflicts, and then pull out the changes into a separate PR. Agreed?

jwnimmer-tri commented:
> ... wait for #2207 to be merged, fix the conflicts, and then pull out the changes into a separate PR. Agreed?

Sounds fine to me.  Whatever's easiest for you.  For reviewers, I think all we care about is that the reviewable.io text is the appropriate changes to inspect.

liangfok commented:
__Note to reviewers:__  All of the files in this PR are ready to be reviewed based on the current state of the master branch. However, there is another PR, https://github.com/RobotLocomotion/drake/pull/2207, that is about to be merged and which will likely introduce conflicts. I do not anticipate the conflicts requiring major code refactoring. Most likely, it will only require small changes like switching from `shared_ptr` to just a raw pointer.

<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> > ... wait for #2207 to be merged, fix the conflicts, and then pull out the changes into a separate PR. Agreed?
> 
> 
> 
> Sounds fine to me.  Whatever's easiest for you.  For reviewers, I think all we care about is that the reviewable.io text is the appropriate changes to inspect.</details>


---

Review status: 0 of 33 files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2253#-:-KHNiZEyrUTpoGj8MkYA:868252388)*
<!-- Sent from Reviewable.io -->

liangfok commented:
__Update__: PR #2207 was merged and, as expected, conflicts emerged with this branch. I will fix the conflicts and update this PR ASAP. I will then split this PR into two.

<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> __Note to reviewers:__  All of the files in this PR are ready to be reviewed based on the current state of the master branch. However, there is another PR, https://github.com/RobotLocomotion/drake/pull/2207, that is about to be merged and which will likely introduce conflicts. I do not anticipate the conflicts requiring major code refactoring. Most likely, it will only require small changes like switching from `shared_ptr` to just a raw pointer.</details>


---

Review status: 0 of 33 files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2253#-:-KHPVPuGkPf3M3aed5OV:-682935343)*
<!-- Sent from Reviewable.io -->

liangfok commented:
__Update__: I've resolved all conflicts with master and thus this PR is ready for review. I also broke out part of this PR into https://github.com/RobotLocomotion/drake/pull/2303, which can be reviewed separately if desired. I look forward to the reviews!

<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> __Update__: PR #2207 was merged and, as expected, conflicts emerged with this branch. I will fix the conflicts and update this PR ASAP. I will then split this PR into two.</details>


---

Review status: 0 of 33 files reviewed at latest revision, 7 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2253#-:-KHQ0nkXtJJghwteAMBi:96945102)*
<!-- Sent from Reviewable.io -->

nkoenig commented:
Reviewed 14 of 30 files at r9, 11 of 17 files at r16, 3 of 4 files at r17.
Review status: 27 of 33 files reviewed at latest revision, 12 unresolved discussions, some commit checks failed.

---

*[drake/examples/Cars/car.h, line 49 \[r17\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2253#-KHWekV4n-YTgV9cjtt9:-KHWekV4n-YTgV9cjttA:-1969544937) ([raw file](https://github.com/robotlocomotion/drake/blob/f0eeeb6a17bd44f81c9cf8b8f2951cbf313e8792/drake/examples/Cars/car.h#L49)):*
> ```objc
>  *
>  * @param[out] sim_options A pointer to where the simulation options should
>  * be saved.
> ```

Might want to add doxygen \throws information.

---

*[drake/examples/Cars/car_sim_lcm_and_ros.cpp, line 50 \[r17\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2253#-KHWfALCytzZSTafe9kW:-KHWfALD_D7pVVHIjrCY:-1314370041) ([raw file](https://github.com/robotlocomotion/drake/blob/f0eeeb6a17bd44f81c9cf8b8f2951cbf313e8792/drake/examples/Cars/car_sim_lcm_and_ros.cpp#L50)):*
> ```cpp
>   // Initializes the simulation options.
>   SimulationOptions options;
>   drake::SetSimulationOptions(&options);
> ```

This can throw. Did you want to wrap this in a try block, or let the exception pass up?

---

*[drake/examples/Cars/README.md, line 64 \[r17\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2253#-KHWhdSPkOeQ4UTzj9ev:-KHWhdSQ_PDKIu4tQco_:1859505604) ([raw file](https://github.com/robotlocomotion/drake/blob/f0eeeb6a17bd44f81c9cf8b8f2951cbf313e8792/drake/examples/Cars/README.md#L64)):*
> ```Markdown
> -------------------------
> 
> There are two version of Drake's cars simulator, one that integrates only
> ```

`version` --> `versions`

---

*[drake/ros/sensor_visualizer_lidar.h, line 173 \[r17\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2253#-KHWqrzELuDCxdUuGz7i:-KHWqrzFS5wxoi0doNLr:670219196) ([raw file](https://github.com/robotlocomotion/drake/blob/f0eeeb6a17bd44f81c9cf8b8f2951cbf313e8792/drake/ros/sensor_visualizer_lidar.h#L173)):*
> ```objc
>     // Checks whether enough time has elapsed since the last transmission.
>     // Aborts if insufficient time has passed. This is to prevent flooding ROS
>     // topic /tf.
> ```

I think you meant to say "prevent flooding ROS sensor topics".

---

*[drake/systems/plants/RigidBody.h, line 22 \[r17\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2253#-KHWsTue9tfsFJWa8NV9:-KHWsTufXn40whfvZiI0:1041068774) ([raw file](https://github.com/robotlocomotion/drake/blob/f0eeeb6a17bd44f81c9cf8b8f2951cbf313e8792/drake/systems/plants/RigidBody.h#L22)):*
> ```objc
>   RigidBody();
> 
>   const std::string& GetName() const;
> ```

Doxygen here and below?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2253)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Dear reviewers,

I merged in the latest from `master` and addressed all reviewer comments. PTAL.

Thank you!

<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> __Update__: I've resolved all conflicts with master and thus this PR is ready for review. I also broke out part of this PR into https://github.com/RobotLocomotion/drake/pull/2303, which can be reviewed separately if desired. I look forward to the reviews!</details>


---

Reviewed 1 of 17 files at r16.
Review status: 19 of 28 files reviewed at latest revision, 12 unresolved discussions.

---

*[drake/examples/Cars/car.h, line 49 \[r17\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2253#-KHWekV4n-YTgV9cjtt9:-KHj78_H3AfiXnrez1iU:1610000531) ([raw file](https://github.com/robotlocomotion/drake/blob/f0eeeb6a17bd44f81c9cf8b8f2951cbf313e8792/drake/examples/Cars/car.h#L49)):*
<details><summary><i>Previously, nkoenig (Nate Koenig) wrote…</i></summary>
> Might want to add doxygen \throws information.</details>
Added.

---

*[drake/examples/Cars/car_sim_lcm_and_ros.cpp, line 50 \[r17\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2253#-KHWfALCytzZSTafe9kW:-KHj870ov4QEN1zGS1FF:1023102047) ([raw file](https://github.com/robotlocomotion/drake/blob/f0eeeb6a17bd44f81c9cf8b8f2951cbf313e8792/drake/examples/Cars/car_sim_lcm_and_ros.cpp#L50)):*
<details><summary><i>Previously, nkoenig (Nate Koenig) wrote…</i></summary>
> This can throw. Did you want to wrap this in a try block, or let the exception pass up?</details>
I've added a `try` / `catch` block around the above method call.
I also added it to the identical code in `car_sim_lcm.cpp`.

---

*[drake/examples/Cars/README.md, line 64 \[r17\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2253#-KHWhdSPkOeQ4UTzj9ev:-KHj93xx18M8DrqVwTrf:-795674652) ([raw file](https://github.com/robotlocomotion/drake/blob/f0eeeb6a17bd44f81c9cf8b8f2951cbf313e8792/drake/examples/Cars/README.md#L64)):*
<details><summary><i>Previously, nkoenig (Nate Koenig) wrote…</i></summary>
> `version` --> `versions`</details>
Fixed. Thanks.

---

*[drake/ros/sensor_visualizer_lidar.h, line 173 \[r17\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2253#-KHWqrzELuDCxdUuGz7i:-KHj612ByspiaPGLKAJM:-795674665) ([raw file](https://github.com/robotlocomotion/drake/blob/f0eeeb6a17bd44f81c9cf8b8f2951cbf313e8792/drake/ros/sensor_visualizer_lidar.h#L173)):*
<details><summary><i>Previously, nkoenig (Nate Koenig) wrote…</i></summary>
> I think you meant to say "prevent flooding ROS sensor topics".</details>
Fixed. Thanks!

---

*[drake/systems/plants/RigidBody.h, line 22 \[r17\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2253#-KHWsTue9tfsFJWa8NV9:-KHj65zfS2lBqI0i2H_2:-2035426538) ([raw file](https://github.com/robotlocomotion/drake/blob/f0eeeb6a17bd44f81c9cf8b8f2951cbf313e8792/drake/systems/plants/RigidBody.h#L22)):*
<details><summary><i>Previously, nkoenig (Nate Koenig) wrote…</i></summary>
> Doxygen here and below?</details>
Fixed. Also, I renamed the methods to be `name()` and `model_name()` to conform to the latest style guide.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2253#-:-KHj9IMg6BmFONcr9A6U:467005276)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Build servers detected problems on Windows. I will look into tye problem soon.

https://drake-cdash.csail.mit.edu/viewBuildError.php?buildid=19948

<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> Dear reviewers,
> 
> I merged in the latest from `master` and addressed all reviewer comments. PTAL.
> 
> Thank you!</details>


---

Review status: 19 of 28 files reviewed at latest revision, 12 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2253#-:-KHk1gM5Rqzv-QWZLKi-:-213324511)*
<!-- Sent from Reviewable.io -->

liangfok commented:
This PR passed [Build #1428](https://drake-jenkins.csail.mit.edu/job/experimental/1428/) and is thus ready for review.

<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> Build servers detected problems on Windows. I will look into tye problem soon.
> 
> https://drake-cdash.csail.mit.edu/viewBuildError.php?buildid=19948</details>


---

Review status: 19 of 28 files reviewed at latest revision, 12 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2253#-:-KHkpndAaJOOOu22dDth:-1788726320)*
<!-- Sent from Reviewable.io -->

david-german-tri commented:
Review status: 19 of 28 files reviewed at latest revision, 12 unresolved discussions.

---

*[drake/examples/Cars/car_sim_lcm_and_ros.cpp, line 50 \[r17\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2253#-KHWfALCytzZSTafe9kW:-KHtaPCELhb9WhMCi06q:55078657) ([raw file](https://github.com/robotlocomotion/drake/blob/f0eeeb6a17bd44f81c9cf8b8f2951cbf313e8792/drake/examples/Cars/car_sim_lcm_and_ros.cpp#L50)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> I've added a `try` / `catch` block around the above method call.
> I also added it to the identical code in `car_sim_lcm.cpp`.</details>
(Jumping in on this particular comment - haven't reviewed anything else)

A few things to consider, no change necessarily needed:

- Writing exception-safe code requires a lot of vigilance.  Right now, the only substantive`catch` statements in the Drake library are for interfacing with NLOPT (plus one thing in a .mex file).  This spares us worrying about whether code is left in a valid state when an exception passes through it.  I'd like to keep things that way.

- However, this isn't library code, it's an executable.  We `throw` on error conditions all the time, which obviously means programs that use Drake might `catch`.  As long as the API contract is "if Drake throws, assume all the Drake objects you were touching are invalid", a `catch` is fine, and even useful.  Your code here is consistent with that contract.

- That said, this particular use doesn't seem to add much value, since all you do with the exception is print an error and exit.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2253)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
I started to add a few comments, and then bailed.  I do not think that we should accept this amount of non-`/dev/` code into the tree unless it has jenkins coverage.  So I think #2116 is a prerequisite for merging this PR as it stands now.

I guess the `Simulation.h` whitespace change could be pulled separately (or reverted).

I will also now go look at `car.h` to see if perhaps that refactoring could also be merged early.

<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> This PR passed [Build #1428](https://drake-jenkins.csail.mit.edu/job/experimental/1428/) and is thus ready for review.</details>


---

Reviewed 1 of 30 files at r9, 2 of 17 files at r16, 3 of 13 files at r18.
Review status: 21 of 28 files reviewed at latest revision, 20 unresolved discussions.

---

*[drake/CMakeLists.txt, line 229 \[r20\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2253#-KHtdLLRr0C22gYNCM6c:-KHtdLLRr0C22gYNCM6d:-1051522039) ([raw file](https://github.com/robotlocomotion/drake/blob/f3c4a180754ec80ddb0eb473b35828914d0e8f29/drake/CMakeLists.txt#L229)):*
> ```CMake
>   # applications that use the custom ROS messages.
>   # add_message_files(DIRECTORY ${CMAKE_CURRENT_SOURCE_DIR}/rostypes FILES
>   #   DrakeFoo.msg)
> ```

Remove commented-out code.

---

*[drake/CMakeLists.txt, line 232 \[r20\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2253#-KHtdNWi98GcTYdpXTbG:-KHtdNWi98GcTYdpXTbH:-1051522039) ([raw file](https://github.com/robotlocomotion/drake/blob/f3c4a180754ec80ddb0eb473b35828914d0e8f29/drake/CMakeLists.txt#L232)):*
> ```CMake
> 
>   # Generates the code for Drake's custom ROS messages.
>   # generate_messages(DEPENDENCIES std_msgs)
> ```

Remove commented-out code.

---

*[drake/CMakeLists.txt, line 235 \[r20\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2253#-KHtdO9_i1RiiOZ_ryGt:-KHtdO9_i1RiiOZ_ryGu:-1051522039) ([raw file](https://github.com/robotlocomotion/drake/blob/f3c4a180754ec80ddb0eb473b35828914d0e8f29/drake/CMakeLists.txt#L235)):*
> ```CMake
> 
>   # Includes the directories containing the generated ROS message code.
>   # include_directories("${${PROJECT_NAME}_INCLUDE_DIRS}")
> ```

Remove commented-out code.

---

*[drake/examples/Cars/ros_packages/drake_cars_examples/CMakeLists.txt, line 1 \[r20\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2253#-KHth0EfHv3T0mhQO09T:-KHth0EfHv3T0mhQO09U:-1615828585) ([raw file](https://github.com/robotlocomotion/drake/blob/f3c4a180754ec80ddb0eb473b35828914d0e8f29/drake/examples/Cars/ros_packages/drake_cars_examples/CMakeLists.txt#L1)):*
> ```CMake
> cmake_minimum_required(VERSION 2.8.3)
> ```

Does this file do anything useful?  It seems like it's all just comments.  And it's not referenced from the Drake build at all, I think?

---

*[drake/examples/Cars/ros_packages/drake_cars_examples/CMakeLists.txt, line 2 \[r20\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2253#-KHtgstbMKHVDKwpSoV_:-KHtgstcovm_nFfGHS4L:354364418) ([raw file](https://github.com/robotlocomotion/drake/blob/f3c4a180754ec80ddb0eb473b35828914d0e8f29/drake/examples/Cars/ros_packages/drake_cars_examples/CMakeLists.txt#L2)):*
> ```CMake
> cmake_minimum_required(VERSION 2.8.3)
> project(drake_cars_examples)
> ```

Shouldn't this be drake_examples_cars to match the directory nesting?

---

*[drake/examples/Cars/test/CMakeLists.txt, line 7 \[r20\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2253#-KHtgBQE-IOdBMqHUaUb:-KHtgBQE-IOdBMqHUaUc:1026449855) ([raw file](https://github.com/robotlocomotion/drake/blob/f3c4a180754ec80ddb0eb473b35828914d0e8f29/drake/examples/Cars/test/CMakeLists.txt#L7)):*
> ```CMake
> endif ()
> 
> if (GTEST_FOUND AND roscpp_FOUND)
> ```

I don't think car_sim_lcm_and_ros depends on GTEST, so doesn't deserve the conditional here?

---

*[drake/examples/Cars/test/CMakeLists.txt, line 9 \[r20\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2253#-KHtftN-Ff15i91_bsvX:-KHtftN0KadZd1FmhTjs:-809855472) ([raw file](https://github.com/robotlocomotion/drake/blob/f3c4a180754ec80ddb0eb473b35828914d0e8f29/drake/examples/Cars/test/CMakeLists.txt#L9)):*
> ```CMake
> if (GTEST_FOUND AND roscpp_FOUND)
>   add_test(NAME car_sim_lcm_and_ros COMMAND car_sim_lcm_and_ros models/prius/prius.sdf models/stata_garage_p1.sdf --duration 5 WORKING_DIRECTORY "${PROJECT_SOURCE_DIR}/examples/Cars")
> endif ()
> ```

Missing newline at end of file.

---

*[drake/ros/CMakeLists.txt, line 8 \[r20\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2253#-KHtdnYTux0f3PcyRSIj:-KHtdnYTux0f3PcyRSIk:-809855472) ([raw file](https://github.com/robotlocomotion/drake/blob/f3c4a180754ec80ddb0eb473b35828914d0e8f29/drake/ros/CMakeLists.txt#L8)):*
> ```CMake
>   drake_install_headers(drake_ros_tf_publisher.h
>                         sensor_visualizer_lidar.h)
> endif()
> ```

Missing newline at end of file.

---

*[drake/ros/drake_ros_tf_publisher.h, line 1 \[r12\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2253#-KGxogLp6VnbWoCXCR17:-KHteEV9eOej8wP7Uk4l:-1007453797) ([raw file](https://github.com/robotlocomotion/drake/blob/7c222f12089e672e34da22832fbd1b6c09c1c278/drake/ros/drake_ros_tf_publisher.h#L1)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> I added unit test `car_sim_lcm_and_ros`, which covers this new system. To run the unit test:
> 
> ```
> $ ctest -VV -R car_sim_lcm_and_ros
> ```</details>
I do not think that code outside of `examples` should be allowed to claim test coverage from code in `examples`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2253#-:-KHtes6hvDbWwxdr7LJ1:1373654831)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
I think the `car.{h,cc}` changes are close enough to being ready that they would make sense as a separate PR that could go in relatively quickly.  (Possibly with a better name though?  `car_simulation.h` or `car_sim.h`; it's not really a car, it's helpers for simulating a car (or multiple cars, once I'd get done refactoring it in my next feature branches.)

<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> I started to add a few comments, and then bailed.  I do not think that we should accept this amount of non-`/dev/` code into the tree unless it has jenkins coverage.  So I think #2116 is a prerequisite for merging this PR as it stands now.
> 
> I guess the `Simulation.h` whitespace change could be pulled separately (or reverted).
> 
> I will also now go look at `car.h` to see if perhaps that refactoring could also be merged early.</details>


---

Review status: 21 of 28 files reviewed at latest revision, 20 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2253#-:-KHts8qmUIJREwElOAph:1221210658)*
<!-- Sent from Reviewable.io -->

liangfok commented:
OK I will pull the `car.{h,cc}` changes into a separate PR. I will probably rename the file to be `car_simulation.h`. Thanks for the advice!

<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> I think the `car.{h,cc}` changes are close enough to being ready that they would make sense as a separate PR that could go in relatively quickly.  (Possibly with a better name though?  `car_simulation.h` or `car_sim.h`; it's not really a car, it's helpers for simulating a car (or multiple cars, once I'd get done refactoring it in my next feature branches.)</details>


---

Review status: 21 of 28 files reviewed at latest revision, 20 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2253#-:-KHu-xbAzDEji2u4_Y_Q:393322136)*
<!-- Sent from Reviewable.io -->

liangfok commented:
I've either addressed or replied to all reviewer comments. I'll next focus on pulling out just the changes to `car_simulation.h` and `car_simulation.cc` into a separate PR.

<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> OK I will pull the `car.{h,cc}` changes into a separate PR. I will probably rename the file to be `car_simulation.h`. Thanks for the advice!</details>


---

Reviewed 1 of 4 files at r14, 2 of 15 files at r18, 1 of 2 files at r20.
Review status: 4 of 31 files reviewed at latest revision, 18 unresolved discussions.

---

*[drake/CMakeLists.txt, line 229 \[r20\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2253#-KHtdLLRr0C22gYNCM6c:-KHuyKV_k_gUQRw6G-Vf:-499146009) ([raw file](https://github.com/robotlocomotion/drake/blob/f3c4a180754ec80ddb0eb473b35828914d0e8f29/drake/CMakeLists.txt#L229)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> Remove commented-out code.</details>
Done.

---

*[drake/CMakeLists.txt, line 232 \[r20\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2253#-KHtdNWi98GcTYdpXTbG:-KHuyL7Z6_vumLaDWf0O:-499146009) ([raw file](https://github.com/robotlocomotion/drake/blob/f3c4a180754ec80ddb0eb473b35828914d0e8f29/drake/CMakeLists.txt#L232)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> Remove commented-out code.</details>
Done.

---

*[drake/CMakeLists.txt, line 235 \[r20\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2253#-KHtdO9_i1RiiOZ_ryGt:-KHuyLxATEAP7DVqwen-:-499146009) ([raw file](https://github.com/robotlocomotion/drake/blob/f3c4a180754ec80ddb0eb473b35828914d0e8f29/drake/CMakeLists.txt#L235)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> Remove commented-out code.</details>
Done.

---

*[drake/examples/Cars/car_sim_lcm_and_ros.cpp, line 50 \[r17\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2253#-KHWfALCytzZSTafe9kW:-KHuyxyX7D-ZehHsnugT:354371095) ([raw file](https://github.com/robotlocomotion/drake/blob/f0eeeb6a17bd44f81c9cf8b8f2951cbf313e8792/drake/examples/Cars/car_sim_lcm_and_ros.cpp#L50)):*
<details><summary><i>Previously, david-german-tri (David German) wrote…</i></summary>
> (Jumping in on this particular comment - haven't reviewed anything else)
> 
> A few things to consider, no change necessarily needed:
> 
> - Writing exception-safe code requires a lot of vigilance.  Right now, the only substantive`catch` statements in the Drake library are for interfacing with NLOPT (plus one thing in a .mex file).  This spares us worrying about whether code is left in a valid state when an exception passes through it.  I'd like to keep things that way.
> 
> - However, this isn't library code, it's an executable.  We `throw` on error conditions all the time, which obviously means programs that use Drake might `catch`.  As long as the API contract is "if Drake throws, assume all the Drake objects you were touching are invalid", a `catch` is fine, and even useful.  Your code here is consistent with that contract.
> 
> - That said, this particular use doesn't seem to add much value, since all you do with the exception is print an error and exit.</details>
OK I modified the exception's text be more informative and removed the tray/catch block.

---

*[drake/examples/Cars/test/CMakeLists.txt, line 7 \[r20\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2253#-KHtgBQE-IOdBMqHUaUb:-KHv0A_WTJBa7DFc9PKf:70658910) ([raw file](https://github.com/robotlocomotion/drake/blob/f3c4a180754ec80ddb0eb473b35828914d0e8f29/drake/examples/Cars/test/CMakeLists.txt#L7)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> I don't think car_sim_lcm_and_ros depends on GTEST, so doesn't deserve the conditional here?</details>
You're right. I've removed it.

---

*[drake/examples/Cars/test/CMakeLists.txt, line 9 \[r20\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2253#-KHtftN-Ff15i91_bsvX:-KHv0HrlFFF5woJa4PmM:-499146009) ([raw file](https://github.com/robotlocomotion/drake/blob/f3c4a180754ec80ddb0eb473b35828914d0e8f29/drake/examples/Cars/test/CMakeLists.txt#L9)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> Missing newline at end of file.</details>
Done.

---

*[drake/ros/CMakeLists.txt, line 8 \[r20\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2253#-KHtdnYTux0f3PcyRSIj:-KHv0MGOUK1fLE3EUNhL:-499146009) ([raw file](https://github.com/robotlocomotion/drake/blob/f3c4a180754ec80ddb0eb473b35828914d0e8f29/drake/ros/CMakeLists.txt#L8)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> Missing newline at end of file.</details>
Done.

---

*[drake/ros/drake_ros_tf_publisher.h, line 1 \[r12\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2253#-KGxogLp6VnbWoCXCR17:-KHvF2-OwhpFsujdnxEO:-984315053) ([raw file](https://github.com/robotlocomotion/drake/blob/7c222f12089e672e34da22832fbd1b6c09c1c278/drake/ros/drake_ros_tf_publisher.h#L1)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> I do not think that code outside of `examples` should be allowed to claim test coverage from code in `examples`.</details>
OK I will create a dedicated unit test for this class.

---

*[drake/examples/Cars/ros_packages/drake_cars_examples/CMakeLists.txt, line 1 \[r20\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2253#-KHth0EfHv3T0mhQO09T:-KHuzJ6znpjnT-eADK7I:-1692852791) ([raw file](https://github.com/robotlocomotion/drake/blob/f3c4a180754ec80ddb0eb473b35828914d0e8f29/drake/examples/Cars/ros_packages/drake_cars_examples/CMakeLists.txt#L1)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> Does this file do anything useful?  It seems like it's all just comments.  And it's not referenced from the Drake build at all, I think?</details>
Yeah it's part of ROS' Catkin build system: http://wiki.ros.org/catkin/CMakeLists.txt

I'm actually wondering whether this ROS package should be in a separate git repository. Thoughts?

---

*[drake/examples/Cars/ros_packages/drake_cars_examples/CMakeLists.txt, line 2 \[r20\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2253#-KHtgstbMKHVDKwpSoV_:-KHuzYHDkMHWtA0jb-W2:-1711452074) ([raw file](https://github.com/robotlocomotion/drake/blob/f3c4a180754ec80ddb0eb473b35828914d0e8f29/drake/examples/Cars/ros_packages/drake_cars_examples/CMakeLists.txt#L2)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> Shouldn't this be drake_examples_cars to match the directory nesting?</details>
No problem. I changed the ROS package name.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2253#-:-KHvF7geYDMRDq32LJeh:950536414)*
<!-- Sent from Reviewable.io -->

liangfok commented:
I am splitting this PR into a number of smaller PRs and will thus close this PR.
